File: android/fed_kit/src/main/java/org/eu/fedcampus/fed_kit/TrainState.kt
package org.eu.fedcampus.fed_kit

import io.grpc.ManagedChannel
import org.eu.fedcampus.fed_kit_train.db.TFLiteModel
import org.eu.fedcampus.fed_kit_train.FlowerClient

sealed class TrainState<X : Any, Y : Any> {
    class Initialized<X : Any, Y : Any> : TrainState<X, Y>()

    data class WithModel<X : Any, Y : Any>(val model: TFLiteModel) : TrainState<X, Y>()

    data class Prepared<X : Any, Y : Any>(
        val model: TFLiteModel,
        val flowerClient: FlowerClient<X, Y>,
        val channel: ManagedChannel
    ) : TrainState<X, Y>()

    data class Training<X : Any, Y : Any>(
        val model: TFLiteModel,
        val flowerClient: FlowerClient<X, Y>,
        val flowerServiceRunnable: FlowerServiceRunnable<X, Y>
    ) : TrainState<X, Y>()
}

File: flutter/lib/train_state.dart
import 'package:fed_kit/flower_service.dart';
import 'package:fed_kit/ml_client.dart';
import 'package:fed_kit/ml_model.dart';
import 'package:grpc/grpc.dart';

sealed class TrainState {}

class Initialized extends TrainState {}

class WithModel extends TrainState {
  final MLModel model;

  WithModel(this.model);
}

/// The platform side has a `FlowerClient` ready.
class Prepared extends TrainState {
  final MLModel model;
  final MLClient mlClient;
  final ClientChannel channel;

  Prepared(this.model, this.mlClient, this.channel);
}

class Training extends TrainState {
  final MLModel model;
  final FlowerService flowerService;

  Training(this.model, this.flowerService);
}
